### ISSUE TITLE
Some feature ideas for 'mini.files'

### ISSUE THREAD

**Issue by IndianBoy42** [echasnovski/mini.nvim#378]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.files

### Description

Sorry to dump this all in one issue

1. LSP file operation notifications (willRenameFile, willCreateFile, willDeleteFile)
    - These methods (may) return workspace edits, seeing the preview of those before hitting synchronize would be useful
2. Rendering the icons using inline virtual text so they don't get in the way when editing buffers.
    - Also could render other file metadata from the filesystem or git or lsp/diagnostics
3. Hooks for controlling where to open the file.
   - Obviously things like `open in split` and `open in tab` could be builtin but having the user able to put in a function would allow using [nvim-window-picker](https://github.com/s1n7ax/nvim-window-picker)
4. Preview for files
5. Auto close on focus lost

**Comment by IndianBoy42** [comment]
> then this situation should not happen with proper use.

Hmm, well I could be holding it wrong since its a new plugin but it happened to me almost immediately, because my mappings for changing windows (as well as mouse). I guess I see the problem if there is some synchronization to do but when using it just to explore and open files I can see users accidentally getting into this situation. 

> >Preview for files

> Out of scope due to implementation complexity.

Isn't it just opening the file (maybe with nobuflisted) and closing it when the cursor moves?

> Rendering any other data is [out of scope](https://github.com/echasnovski/mini.nvim/blob/4b0881b9fa22ddce26ba30cdf3711226e89ba81f/doc/mini-files.txt#L32).

I see, I'll take a crack at it as an extension and maybe it could be a part of a 'cookbook' of recipes for sharing would be nice because I can definitely see a lot of people wanting it. 

> I'll think about adding dedicated autocommand events for every file system action, but this will be as much as it will get in 'mini.files'.

I think this would be sufficient and very valuable, may need events before and after actually doing the file operation.

> Inline virtual text is not even in stable Neovim version, so this probably won't happen for a while.

True, and if you're not up for gating it behind a version check I'll just wait, but I've already encountered a problem that if the icon is accidentally backspaced away then mini.files interprets it as a glitched rename (`RENAME: 'lsp-file-operations.lua' to '07lsp-file-operations.lua'`)

**Comment by echasnovski** [comment]
> Hmm, well I could be holding it wrong since its a new plugin but it happened to me almost immediately, because my mappings for changing windows (as well as mouse). I guess I see the problem if there is some synchronization to do but when using it just to explore and open files I can see users accidentally getting into this situation.

Ah, the mouse can mess things, yes. Otherwise, focus should not be lost if doing `h`/`j`/`k`/`l` or in-buffer movement. But I'll think about it.

> Isn't it just opening the file (maybe with nobuflisted) and closing it when the cursor moves?

It is not that easy, unfortunately. There are also issue of how to properly create and delete buffer for the file; track it inside explorer object for proper history restore (and now buffers are only designed to show directories); deal with `FileType` autocommands, LSPs, etc.

> I think this would be sufficient and very valuable, may need events before and after actually doing the file operation.

I am thinking only about adding event *after* successful operation. What are the use cases for *before* it?

> True, and if you're not up for gating it behind a version check I'll just wait, but I've already encountered a problem that if the icon is accidentally backspaced away then mini.files interprets it as a glitched rename (`RENAME: 'lsp-file-operations.lua' to '07lsp-file-operations.lua'`)

I am more worried about feature stability and line of code in 'mini.files'. It is already in the top 3 of the biggest modules with ambitions to become the top, so I have to be careful now.

Besides, if there was no icon, accidental backspace will alter path index before it, which is *much* worse.

**Comment by IndianBoy42** [comment]
> I am thinking only about adding event after successful operation. What are the use cases for before it?

Well, I'll have to look into it closer but it is called *will*RenameFile, and actually *did*RenameFile is a separate server method. I don't know how much it affects exactly

**Comment by AllenDang** [comment]
+1 for Hooks for controlling where to open the file.

**Comment by heygarrett** [comment]
> Hooks for controlling where to open the file.

I was able to create vertical split keymap using `get_fs_entry`:
```lua
return {
	"https://github.com/echasnovski/mini.files",
	config = function()
		local mini_files = require("mini.files")

		local window_handle
		vim.api.nvim_create_user_command("Files", function()
			window_handle = vim.api.nvim_get_current_win()
			mini_files.open(vim.api.nvim_buf_get_name(0))
		end, { desc = "open mini.files browser" })

		local group = vim.api.nvim_create_augroup("mini.files", { clear = true })
		vim.api.nvim_create_autocmd("User", {
			desc = "define commands/keymaps for mini.files",
			pattern = "MiniFilesBufferCreate",
			group = group,
			callback = function(args)
				local buf_id = args.data.buf_id
				vim.keymap.set("n", "<c-v>", function()
					local entry = mini_files.get_fs_entry()
					if not entry then
						return
					end
					vim.api.nvim_win_call(
						window_handle,
						function()
							vim.cmd.vsplit({
								args = { entry.path },
							})
						end
					)
				end, {
					buffer = buf_id,
					desc = "open from mini.files in vertical split",
				})
			end,
		})
	end,
}
```

**Comment by IndianBoy42** [comment]
Not exactly what I thought but I think it could work. I would want the window picker to open when I hit `l/L` on the file rather than needed a separate key sequence, which is why I was thinking of a hook. But I can use the get_fs_entry method along with set_target_window, so it should be fine.

> Icons are probably never going to be implemented as inline extmarks. The main reason is the icons should disappear when line is deleted, which won't be the case with extmarks.

Well they may not immediately disappear but of course you could update the extmarks. Its fine, it was just a suggestion for solving the issues around deleting the icon through accident backspacing or anything.

I really hope to see file operation events, it unlocks a lot of configurability

**Comment by echasnovski** [comment]
I've just pushed update with events triggered after successful file system actions. Here is [a documentation](https://github.com/echasnovski/mini.nvim/blob/445f73cff53987432cb9eea41c4dc25e4a4c3c04/doc/mini-files.txt#L321-L338). I suspect this is as much as I am willing to compromise for adding LSP related file action stuff.

With this, points 1-4 I consider complete.

I am still thinking about closing after losing focus.

**Comment by heygarrett** [comment]
My use case for closing after losing focus:
1. `nvim .` in project root
2. Open Telescope to search for something

Currently, after step 1 mini.files is open, but after step 2 mini.files is still open and Telescope is open underneath it.

**Comment by echasnovski** [comment]
@heygarrett, is there a reason you open Neovim with `nvim .` even if you don't want to look at directory content inside Neovim? Plain `nvim` should still open Neovim with correct current working directory.

**Comment by heygarrett** [comment]
@echasnovski Apologies, that was a contrived example. In actuality I have a sessions workflow that defaults to opening my file browser if there's no session and no argument was provided to `nvim`.

Now that I think about it though, I suppose I'm just not accustomed to having my file browser in a floating window; I was previously opening Telescope or FTerm over oil.nvim without conflict.

**Comment by IndianBoy42** [comment]
Actually could we just ourselves create BufLeave or WinLeave autocmds from the MiniFilesBufferCreate autocmd to do `quit on focus lost` ourselves?

**Comment by heygarrett** [comment]
@IndianBoy42 Good idea! Here's a, perhaps unorthodox, solution. 😅

```lua
local group = vim.api.nvim_create_augroup("mini.files", { clear = true })
vim.api.nvim_create_autocmd("User", {
	pattern = "MiniFilesBufferCreate",
	desc = "define commands/keymaps for mini.files",
	group = group,
	callback = function(args)
		local buf_id = args.data.buf_id
		vim.api.nvim_create_autocmd("WinLeave", {
			group = group,
			buffer = buf_id,
			callback = function()
				vim.api.nvim_create_autocmd("BufEnter", {
					group = group,
					once = true,
					callback = function()
						if vim.bo.filetype ~= "minifiles" then
							vim.api.nvim_win_call(0, require("mini.files").close)
						end
					end,
				})
			end,
		})
	end,
})
```

**Comment by echasnovski** [comment]
I did plan to support "close on lost focus", but it has some small technical difficulties during testing. In short: it autocommand callback needs to be wrapped in `schedule_wrap()` to allow scripts to temporary change windows; but this leads to an unstable testing.

The real problem here is that any current ad-hoc solutions will virtually disable `show_help` functionality (as its buffer is not 'minifiles' filetype). I am planning tomorrow to push update where help window will have separate 'minifiles-help' filetype and `close()` will close help window too. After it is active, I'd suggest using this autocommand for "close on lost focus":

```lua
-- Use `vim.schedule_wrap()` to allow temporary "fast" window chang
local minifiles_track_lost_focus = vim.schedule_wrap(function()
  local is_in_focus = vim.bo.filetype == 'minifiles' or vim.bo.filetype == 'minifiles-help'
  if is_in_focus then return end
  MiniFiles.close()
end)

vim.api.nvim_create_autocmd(
  'BufEnter',
  { callback = minifiles_track_lost_focus, desc = "Close 'mini.files' on lost focus" }
)
```

**Comment by echasnovski** [comment]
Alright, I've just pushed a big-ish update to 'mini.files' with final pieces addressing this issue.

- Files are now previewed after 8744876fd48e5be195379cd5fbc6b8f8e08f439c .
- Auto close on focus lost will not be built-in, but it is made possible with manual autocommand. See previous post.

Closing this, as I think everything from this issue is now addressed.

**Comment by echasnovski** [comment]
> @echasnovski I've installed the update and tested it with the autocommand code you wrote above, but unfortunately it doesn't work when an FZF floating window buffer is launched (minifile is still open with focus and the FZF window is opened behind it). If instead I remove the `vim.schedule_wrap`, it works in that minifile is correctly closed and the FZF window has the cursor focus, however if the FZF window is immediately closed the cursor doesn't return to the original buffer from which minifiles was opened (in my case the cursor is left in the satellite.nvim scroll bar)

How do you open "FZF floating window buffer"? I would suspect that opening it would end up with it being in focus (its buffer and window being current) and not any 'mini.files' buffer.

**Comment by echasnovski** [comment]
> I use the command `:Files` (or the keybinding I've mapped to this command `C-f`). When minifiles keeps the focus (and the FZF floating window is below it in z-height), and then minifiles is closed with `q`, the focus goes back to the original buffer window rather than the floating FZF window so I'm not sure if the FZF window ever had focus

Oh, I see. That's right, because `MiniFiles.close()` indeed forces focus on its target window afterwards.

I've also just tried with `:Telescope` and it closes it along with 'mini.files' (I guess because 'telescope.nvim' itself has some strict "close on lost focus" logic).

My best suggestion here right now would then be to save and restore proper current window:
```lua
-- Use `vim.schedule_wrap()` to allow temporary "fast" window chang
local minifiles_track_lost_focus = vim.schedule_wrap(function()
  local ft = vim.bo.filetype
  if ft == 'minifiles' or ft == 'minifiles-help' then return end
  local cur_win_id = vim.api.nvim_get_current_win()
  MiniFiles.close()
  pcall(vim.api.nvim_set_current_win, cur_win_id)
end)

vim.api.nvim_create_autocmd(
  'BufEnter',
  { callback = minifiles_track_lost_focus, desc = "Close 'mini.files' on lost focus" }
)
```

Otherwise, closing explorer prior to some conscious focusing on other buffer is the best approach here. This autocommand is more targeted towards "close explorer if clicked with mouse outside of it".

**Comment by tomspeak** [comment]
> There is also a [new example](https://github.com/echasnovski/mini.nvim/blob/52a5f78c6b24bfd76b2c7baa66a0061a4635d6e5/doc/mini-files.txt#L367) for suggested way to handle "open file in a split" use case. The way it works is more granular than in 'oil.nvim': first create split from within 'mini.files' (`gv` for vertical split, `gs` for horizontal split) and then open file.

Is there any chance the behavior could somehow be opening into a split in one movement? It's a small nit but I am struggling to get used to split then open into split.

**Comment by Armadillidiid** [comment]
Can we have the ability to scroll preview with `<c-f>` and `<c-b>` just like mini.pick`? Some files are really long and don't want to open the file as a buffer.

**Comment by echasnovski** [comment]
> Can we have the ability to scroll preview with `<c-f>` and `<c-b>` just like mini.pick`? Some files are really long and don't want to open the file as a buffer.

I am afraid, that is probably out of scope. The reason for this is that absent scrolling allows to read not the whole file/directory to preview it. Which is important for performance.

To be fair, the 'mini.pick' also has this kind of optimization (it loads content enough for several scrolls). But the preview in 'mini.files' is not meant to actually look at the content. The main goal is to give a quick glance at whether the file is the one user thinks it is.
